test/toolkits/style_transfer/utils.mm

```diff
@@ -57,29 +71,1189 @@
+
+  nn_spec->add_inner_product(
+      /* name */ "transformer_instancenorm5_beta",
+      /* input */ "index",
+      /* num_output_channels */ 3,
+      /* num_input_channels */ 8,
+      /* weight_init_fn */ zero_weight_initializer(),
+      /* bias_init_fn */ zero_weight_initializer());
+
+  nn_spec->add_instancenorm(
+      /* name */ "transformer_instancenorm5__fwd_bn_",
+      /* input */ "transformer_conv5",
+      /* num_channels */ 3,
+      /* epsilon */ 9.99999974738e-06);
+
+  nn_spec->add_multiplication(
+      /* name */ "transformer_instancenorm5__fwd_mult_gamma",
+      /* inputs */ {"transformer_instancenorm5__fwd_bn_", "transformer_instancenorm5_gamma"});
+
+  nn_spec->add_addition(
+      /* name */ "transformer_instancenorm5__fwd",
+      /* inputs */ {"transformer_instancenorm5__fwd_mult_gamma", "transformer_instancenorm5_beta"});
+
+  nn_spec->add_sigmoid(
+      /* name */ "transformer_activation5",
+      /* input */ "transformer_instancenorm5__fwd");
+
+  nn_spec->add_scale(
+      /* name */ "stylizedImage",
+      /* input */ "transformer_activation5",
+      /* shape_c_h_w */ {1, 1, 1},
+      /* weight_init_fn */ scalar_weight_initializer(255.0));
+
+  const std::string TRANSFORMER_NAME = "./transformer.mlmodel";
+
+  CoreML::Specification::Model model;
+  model.set_specificationversion(3);
+
+  // Set Input Image
+  ModelDescription* model_desc = model.mutable_description();
+
+  FeatureDescription* input_image_feature_desc = model_desc->add_input();
+  input_image_feature_desc->set_name("image");
+  input_image_feature_desc->set_shortdescription("Input image");
+
+  ImageFeatureType* input_image_feature =
+      input_image_feature_desc->mutable_type()->mutable_imagetype();
+  input_image_feature->set_width(256);
+  input_image_feature->set_height(256);
+  input_image_feature->set_colorspace(ImageFeatureType::RGB);
+
+  // Set Input Index
+  FeatureDescription* input_index_feature_desc = model_desc->add_input();
+  input_index_feature_desc->set_name("index");
+  input_index_feature_desc->set_shortdescription(
+      "Style index array (set index I to 1.0 to enable Ith style)");
+
+  ArrayFeatureType* array = input_index_feature_desc->mutable_type()->mutable_multiarraytype();
+  array->add_shape(1);
+  array->set_datatype(ArrayFeatureType::DOUBLE);
+
+  // Set Output
+  FeatureDescription* output_feature_desc = model_desc->add_output();
+  output_feature_desc->set_name("stylizedImage");
+  input_index_feature_desc->set_shortdescription("Stylized image");
+
+  ImageFeatureType* output_image_feature = output_feature_desc->mutable_type()->mutable_imagetype();
+  output_image_feature->set_width(256);
+  output_image_feature->set_height(256);
+  output_image_feature->set_colorspace(ImageFeatureType::RGB);
+
+  model.mutable_neuralnetwork()->MergeFrom(nn_spec->get_coreml_spec());
+
+  auto model_wrapper = std::make_shared<MLModelWrapper>(std::make_shared<CoreML::Model>(model));
```

Comment:
Probably shouldn't use `auto` here.

Reply:
Honestly, I think this is one of the cases where `auto` makes sense: `std::make_shared<T>` _always_ returns a value of type `std::shared_ptr<T>`, so the reader can tell from local information (that line alone) what `auto` resolves too. (See some discussion here: https://google.github.io/styleguide/cppguide.html#Type_deduction)

My rule of thumb is that `auto` is okay if the type can be exactly deduced from immediate context, or if it saves at least one level of nested STL type nonsense. e.g. for
```
  std::map<std::string,std::vector<int>> m;
  for (const auto& kv : m) { // prefer `auto` over `std::map<std::string,std::vector<int>>::value_type` or `std::pair<std::string, std::vector<int>>`
```
but YMMV

Reply:
Thanks for the explanation @nickjong.